### DISCUSSION TITLE
[Possible bug] Rendering problems over ssh (backspace, clear).

### DISCUSSION THREAD

**Discussion by blychs** [ghostty-org/ghostty#3144] [Q&A] [answered]
Some of this might be a problem in my system. Since if I type "`clear`", I get
`'xterm-ghostty': unknown terminal type`, and the workaround of doing 
`export TERM=xterm`
 seems to fix everything. I find it surprising, though, since I wouldn't think that anybody added support for, e.g., Kitty explicitly.
If you believe that this is a problem with the remote server itself, feel free to close this issue, but please give it a read and if there is a better workaround, let me know.

Testing Ghostty (thanks for the great work!), I came across the following issue:
The rendering of the backspace/clear over ssh is wrong. This is not the real functionality (which is working fine), but the rendering. Backspace is, in my system, rendered as a forward space. **This problem does not happen when I run Ghostty in my local computer**. It does not happen over other terminal emulators (tested Kitty, ITerm2, xterm).
Local computer: MacPro 14 inch, Apple M2 Pro, MacOS: 14.6.1

ssh HPC system: Derecho, managed by CISL, running suse.

```
plichtig@derecho3:~> cat /etc/os-release
NAME="SLES"
VERSION="15-SP4"
VERSION_ID="15.4"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP4"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp4"
DOCUMENTATION_URL="https://documentation.suse.com/"
plichtig@derecho3:~>
```

Some renders of the problem:
If I type "HiGhostty"

<img width="277" alt="Image" src="https://github.com/user-attachments/assets/046a7be4-20d8-41a9-ac2e-a68c2c3f5e83" />

If I type backspace twice (the cursor moves forward instead of backwards)
<img width="277" alt="Image" src="https://github.com/user-attachments/assets/2855abd9-e376-4fa9-a072-1fedb74ad365" />

However, the characers were actually deleted, and the problem seems to be rendering. If I press Enter, I get (cnf is a search system implmented in our HPC)
<img width="861" alt="Image" src="https://github.com/user-attachments/assets/7cb19454-43a9-4400-af7e-92b5e5381134" />

**Comment by mitchellh** [comment, marked as answer]
https://ghostty.org/docs/help/terminfo

**Comment by blychs** [comment]
This completely answers the issue. I am sorry for not having seen it in the docs. Thanks for the great work!

**Comment by theuves** [comment]
I just got the same error here (on a SSH connection). Thanks.

**Reply by blychs** [reply]
@theuves After logging off and on again, it seems to be reading the xterm-ghostty in .terminfo

**Comment by ickymettle** [comment]
Curious on the tradeoff chosen here - i'm assuming you're eventually wanting to develop a more modern terminfo config for ghostty which is a great idea however given a lot of the user base are likely to spend a lot of time ssh'ing into other machines the new user experience isn't great. Where installing ghostty then ssh'ing into a machine gives you a broken terminal.

Having spent 25+ years on UNIX terminals I had a hunch and knew where to look, for a lot of newer players who haven't had to deal with termcap/terminfo issues over the years it may not be immediately obvious. Maybe one to rethink in future releases?

**Reply by mitchellh** [reply]
It's strictly correct to have a new terminfo because we're a different terminal. 

But beyond that, we support features beyond say `xterm-256color` (and don't support some other legacy features). So, it'd be incorrect to use that as well and break other things. For example, we advertise the Kitty keyboard protocol, synchronized rendering, and more. If you don't use Ghostty's terminfo, some programs "break" (degrade) to not use those features.

**Comment by sviperz** [comment]
> https://ghostty.org/docs/help/terminfo#ssh

When I'm trying the suggested command, I'm getting the following output:

```sh
➜  ~ infocmp -x | ssh 10.112.88.74 -- tic -x -
#       Reconstructed via infocmp from file: /usr/share/terminfo/x/xterm-ghostty
xterm-ghostty|ghostty|Ghostty,
        am, bce, ccc, hs, km, mc5i, mir, msgr, npc, xenl, AX, Su, Tc, XT, fullkbd,
        colors#0x100, cols#80, it#8, lines#24, pairs#0x7fff,
        acsc=++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
        bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
        clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
        csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
        cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
        cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
        cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
        dl=\E[%p1%dM, dl1=\E[M, dsl=\E]2;\007, ech=\E[%p1%dX,
        ed=\E[J, el=\E[K, el1=\E[1K, flash=\E[?5h$<100/>\E[?5l,
        fsl=^G, home=\E[H, hpa=\E[%i%p1%dG, ht=^I, hts=\EH,
        ich=\E[%p1%d@, ich1=\E[@, il=\E[%p1%dL, il1=\E[L, ind=\n,
        indn=\E[%p1%dS,
        initc=\E]4;%p1%d;rgb:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
        invis=\E[8m, kDC=\E[3;2~, kEND=\E[1;2F, kHOM=\E[1;2H,
        kIC=\E[2;2~, kLFT=\E[1;2D, kNXT=\E[6;2~, kPRV=\E[5;2~,
        kRIT=\E[1;2C, kbs=^?, kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB,
        kcuf1=\EOC, kcuu1=\EOA, kdch1=\E[3~, kend=\EOF, kent=\EOM,
        kf1=\EOP, kf10=\E[21~, kf11=\E[23~, kf12=\E[24~,
        kf13=\E[1;2P, kf14=\E[1;2Q, kf15=\E[1;2R, kf16=\E[1;2S,
        kf17=\E[15;2~, kf18=\E[17;2~, kf19=\E[18;2~, kf2=\EOQ,
        kf20=\E[19;2~, kf21=\E[20;2~, kf22=\E[21;2~,
        kf23=\E[23;2~, kf24=\E[24;2~, kf25=\E[1;5P, kf26=\E[1;5Q,
        kf27=\E[1;5R, kf28=\E[1;5S, kf29=\E[15;5~, kf3=\EOR,
        kf30=\E[17;5~, kf31=\E[18;5~, kf32=\E[19;5~,
        kf33=\E[20;5~, kf34=\E[21;5~, kf35=\E[23;5~,
        kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q, kf39=\E[1;6R,
        kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~, kf42=\E[17;6~,
        kf43=\E[18;6~, kf44=\E[19;6~, kf45=\E[20;6~,
        kf46=\E[21;6~, kf47=\E[23;6~, kf48=\E[24;6~,
        kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q, kf51=\E[1;3R,
        kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
        kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
        kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
        kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
        kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
        kind=\E[1;2B, kmous=\E[<, knp=\E[6~, kpp=\E[5~,
        kri=\E[1;2A, oc=\E]104\007, op=\E[39;49m, rc=\E8,
        rep=%p1%c\E[%p2%{1}%-%db, rev=\E[7m, ri=\EM,
        rin=\E[%p1%dT, ritm=\E[23m, rmacs=\E(B, rmam=\E[?7l,
        rmcup=\E[?1049l, rmir=\E[4l, rmkx=\E[?1l\E>, rmso=\E[27m,
        rmul=\E[24m, rs1=\E]\E\\\Ec, sc=\E7,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
        sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
        sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
        smcup=\E[?1049h, smir=\E[4h, smkx=\E[?1h\E=, smso=\E[7m,
        smul=\E[4m, tbc=\E[3g, tsl=\E]2;, u6=\E[%i%d;%dR, u7=\E[6n,
        u8=\E[?%[;0123456789]c, u9=\E[c, vpa=\E[%i%p1%dd,
        BD=\E[?2004l, BE=\E[?2004h, Clmg=\E[s,
        Cmg=\E[%i%p1%d;%p2%ds, Dsmg=\E[?69l, E3=\E[3J,
        Enmg=\E[?69h, Ms=\E]52;%p1%s;%p2%s\007, PE=\E[201~,
        PS=\E[200~, RV=\E[>c, Se=\E[2 q,
        Setulc=\E[58:2::%p1%{65536}%/%d:%p1%{256}%/%{255}%&%d:%p1%{255}%&%d%;m,
        Smulx=\E[4:%p1%dm, Ss=\E[%p1%d q,
        Sync=\E[?2026%?%p1%{1}%-%tl%eh%;,
        XM=\E[?1006;1000%?%p1%{1}%=%th%el%;, XR=\E[>0q,
        fd=\E[?1004l, fe=\E[?1004h, kDC3=\E[3;3~, kDC4=\E[3;4~,
        kDC5=\E[3;5~, kDC6=\E[3;6~, kDC7=\E[3;7~, kDN=\E[1;2B,
        kDN3=\E[1;3B, kDN4=\E[1;4B, kDN5=\E[1;5B, kDN6=\E[1;6B,
        kDN7=\E[1;7B, kEND3=\E[1;3F, kEND4=\E[1;4F,
        kEND5=\E[1;5F, kEND6=\E[1;6F, kEND7=\E[1;7F,
        kHOM3=\E[1;3H, kHOM4=\E[1;4H, kHOM5=\E[1;5H,
        kHOM6=\E[1;6H, kHOM7=\E[1;7H, kIC3=\E[2;3~, kIC4=\E[2;4~,
        kIC5=\E[2;5~, kIC6=\E[2;6~, kIC7=\E[2;7~, kLFT3=\E[1;3D,
        kLFT4=\E[1;4D, kLFT5=\E[1;5D, kLFT6=\E[1;6D,
        kLFT7=\E[1;7D, kNXT3=\E[6;3~, kNXT4=\E[6;4~,
        kNXT5=\E[6;5~, kNXT6=\E[6;6~, kNXT7=\E[6;7~,
        kPRV3=\E[5;3~, kPRV4=\E[5;4~, kPRV5=\E[5;5~,
        kPRV6=\E[5;6~, kPRV7=\E[5;7~, kRIT3=\E[1;3C,
        kRIT4=\E[1;4C, kRIT5=\E[1;5C, kRIT6=\E[1;6C,
        kRIT7=\E[1;7C, kUP=\E[1;2A, kUP3=\E[1;3A, kUP4=\E[1;4A,
        kUP5=\E[1;5A, kUP6=\E[1;6A, kUP7=\E[1;7A, kxIN=\E[I,
sviperz@10.112.88.74's password: rv=\E\\[[0-
Permission denied, please try again.
sviperz@10.112.88.74's password:
```

It doesn't allow me to enter the password, but inserts a part of `infocmp -x` output as a password and then the terminal hangs out completely so I can't enter the password anyway. What am I missing?

**Comment by armetiz** [comment]
I'm saying the above from a UX point of view. 

I had the problem 5 minutes after installing ghosty, and was curious enough to check github to see if it had been discussed. Without this information I would have uninstalled ghostty. 

On the documentation, it's noted in one of the first sentences: ‘Install Ghostty and run! Zero configuration required to get up and running." Which is not true ;)

BTW, the general feeling makes me want to keep using it, good job! 👍

**Reply by RocketLR** [reply]
Exactly "Zero Config" is kind of misleading and tautological. Like what terminal requires configuration out of the box really? 
Its absolutely useless and a major headache to write export TERM=xterm on every system each tim i SSH into them... just because im using this ghostty terminal.. 

Im kind of annoyed that they didnt include this detail in their slogans. "Install Ghostty and run AND also every time you ssh into a new linux server, append "export TERM=xterm" to the .bashrc file or just create the env variable every damn time you ssh via the ghostty terminal" 
Would have saved me the trouble..

**Reply by mitchellh** [reply]
See my response here: https://github.com/ghostty-org/ghostty/discussions/3161#discussioncomment-11803314
